pkg/hubble/relay/peer.go

```diff
@@ -40,112 +41,150 @@ func newPeerClient(target string, dialTimeout time.Duration) (peerpb.PeerClient,
 	return peerpb.NewPeerClient(conn), conn, nil
 }
 
-func (s *Server) syncPeers() {
-	ctx, cancel := context.WithCancel(context.Background())
-connect:
-	for {
-		cl, conn, err := newPeerClient(s.opts.HubbleTarget, s.opts.DialTimeout)
-		if err != nil {
-			s.log.WithFields(logrus.Fields{
-				"error":        err,
-				"dial timeout": s.opts.DialTimeout,
-			}).Warning("Failed to create peer client for peers synchronization; will try again after the timeout has expired")
-			select {
-			case <-s.stop:
-				cancel()
-				return
-			case <-time.After(s.opts.RetryTimeout):
-				continue
-			}
-		}
-		client, err := cl.Notify(ctx, &peerpb.NotifyRequest{})
-		if err != nil {
-			conn.Close()
-			s.log.WithFields(logrus.Fields{
-				"error":              err,
-				"connection timeout": s.opts.RetryTimeout,
-			}).Warning("Failed to create peer notify client for peers change notification; will try again after the timeout has expired")
-			select {
-			case <-s.stop:
-				cancel()
-				return
-			case <-time.After(s.opts.RetryTimeout):
-				continue
-			}
-		}
+type peerSyncer interface {
+	Start()
+	Stop()
+	List() []hubblePeer
+	ConnectPeer(name, address string)
```

Comment:
interface is private, yet fields are public. what's the logic there?

Reply:
This was part of the first refactoring commit, I wasn't sure at the time what the right interface should be. The corresponding interface is now and everything is public because it's in its own package (`pkg/hubble/peer/pool`):
```go
// PeerManager defines the functions a peer manager must implement when
// handling peers and respective connections.
type PeerManager interface {
	// Start instructs the manager to start peer change notification handling
	// and connection management.
	Start()
	// Stop stops any peer manager activity.
	Stop()
	// List returns a list of peers with active connections. If a peer cannot
	// be connected to; its Conn attribute must be nil.
	List() []Peer
	// ReportOffline allows the caller to report a peer as being offline. The
	// peer is identified by its name.
	ReportOffline(name string)
}
```